File: app/src/main/java/idv/markkuo/cscblebridge/MainActivity.java
package idv.markkuo.cscblebridge;

import androidx.appcompat.app.AppCompatActivity;

import android.annotation.SuppressLint;
import android.app.ActivityManager;
import android.content.BroadcastReceiver;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.ServiceConnection;
import android.os.Build;
import android.os.Bundle;
import android.os.IBinder;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

import java.time.Instant;
import java.time.LocalDateTime;
import java.time.ZoneId;


public class MainActivity extends AppCompatActivity  {
    private static final String TAG = MainActivity.class.getSimpleName();

    private TextView tv_speedSensorState, tv_cadenceSensorState, tv_hrSensorState, tv_runSensorState,
            tv_speedSensorTimestamp, tv_cadenceSensorTimestamp, tv_hrSensorTimestamp,
            tv_runSensorTimestamp, tv_speed, tv_cadence, tv_hr, tv_runSpeed, tv_runCadence, tv_time;
    private Button btn_service;

    private boolean serviceStarted = false;
    private MainActivityReceiver receiver;
    private boolean mBound = false;
    private CSCService mService;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        tv_speed = findViewById(R.id.SpeedText);
        tv_cadence = findViewById(R.id.CadenceText);
        tv_time = findViewById(R.id.TimeText);
        btn_service = findViewById(R.id.ServiceButton);

        if (isServiceRunning()) {
            Log.w(TAG, "Service already started");
            serviceStarted = true;
        }

        btn_service.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent mServiceIntent = new Intent(getApplicationContext(), CSCService.class);
                if (!serviceStarted) {
                    Log.d(TAG, "Starting Service");
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O)
                        MainActivity.this.startForegroundService(mServiceIntent);
                    else
                        MainActivity.this.startService(mServiceIntent);

                    // Bind to the service so we can interact with it
                    if (!bindService(mServiceIntent, connection, Context.BIND_AUTO_CREATE)) {
                        Log.d(TAG, "Failed to bind to service");
                    } else {
                        mBound = true;
                    }
                }
                else {
                    Log.d(TAG, "Stopping Service");
                    unbindService();
                    MainActivity.this.stopService(mServiceIntent);
                }

                serviceStarted = !serviceStarted;
                updateButtonState();
            }
        });

//        // Bind to the Sensor title and install a long click listener to restart searching for Speed sensor
//        findViewById(R.id.SpeedSensorText).setOnLongClickListener(createLongClickListener(() -> mService.startSpeedSensorSearch()));
//
//        // Bind to the Sensor title and install a long click listener to restart searching for Cadence sensor
//        findViewById(R.id.CadenceSensorText).setOnLongClickListener(createLongClickListener(() -> mService.startCadenceSensorSearch()));
//
//        // Bind to the Sensor title and install a long click listener to restart searching for HR sensor
//        findViewById(R.id.HRSensorText).setOnLongClickListener(createLongClickListener(() -> mService.startHRSensorSearch()));

        receiver = new MainActivityReceiver();
        // register intent from our service
        IntentFilter filter = new IntentFilter();
        filter.addAction("idv.markkuo.cscblebridge.ANTDATA");
        registerReceiver(receiver, filter);
    }

    // Allows for lambdas on the long click handler to reduce code duplication.
    // TODO: Investigate to see if there is an existing interface we can use
    interface DoAction {
        void action();
    }

    // Long click handler for restarting scan for Ant+ devices
    private View.OnLongClickListener createLongClickListener(final DoAction action) {
        return new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View v) {
                if (!serviceStarted || !mBound) {
                    return false;
                }

                action.action();
                return true;
            }
        };
    }

    private void updateButtonState() {
        // update the title
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                resetUi();
                if (serviceStarted)
                    btn_service.setText(getText(R.string.stop_service));
                else
                    btn_service.setText(getText(R.string.start_service));
            }
        });
    }

    /** Defines callbacks for service binding, passed to bindService() */
    private ServiceConnection connection = new ServiceConnection() {

        @Override
        public void onServiceConnected(ComponentName className,  IBinder service) {
            CSCService.LocalBinder binder = (CSCService.LocalBinder) service;
            mService = binder.getService();
            mBound = true;
        }

        @Override
        public void onServiceDisconnected(ComponentName arg0) {
            mBound = false;
        }
    };

    @Override
    protected void onResume() {
        super.onResume();
        Log.d(TAG, "onResume(): service was " + (serviceStarted ? "started" : "stopped"));
        serviceStarted = isServiceRunning();
    }

    // Unbind from the service
    void unbindService() {
        if (mBound) {
            unbindService(connection);
            mBound = false;
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();

        unregisterReceiver(receiver);
        unbindService();
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
    }

    @Override
    public void onRestoreInstanceState(Bundle savedInstanceState) {
        //resetUi();
        if (serviceStarted)
            btn_service.setText(getText(R.string.stop_service));
        else
            btn_service.setText(getText(R.string.start_service));
    }

    private void resetUi() {
        tv_speed.setText(getText(R.string.no_data));
        tv_cadence.setText(getText(R.string.no_data));
        tv_time.setText("--:--");
    }

    private boolean isServiceRunning() {
        ActivityManager manager = (ActivityManager) getSystemService(Context.ACTIVITY_SERVICE);
        assert manager != null;
        for (ActivityManager.RunningServiceInfo service : manager.getRunningServices(Integer.MAX_VALUE)) {
            if (CSCService.class.getName().equals(service.service.getClassName())) {
                return true;
            }
        }
        return false;
    }

    // this BroadcastReceiver is used to update UI only
    private class MainActivityReceiver extends BroadcastReceiver {
        @Override
        public void onReceive(Context context, Intent intent) {
            final String statusBSD = intent.getStringExtra("bsd_service_status"); // bicycle speed
            final String statusBC = intent.getStringExtra("bc_service_status"); // bicycle cadence
//            final String statusHR = intent.getStringExtra("hr_service_status");
//            final String statusRsc = intent.getStringExtra("ss_service_status");
//            final long speedTimestamp = intent.getLongExtra("speed_timestamp", -1);
//            final long cadenceTimestamp = intent.getLongExtra("cadence_timestamp", -1);
//            final long hrTimestamp = intent.getLongExtra("hr_timestamp", -1);
//            final long runTimestamp = intent.getLongExtra("ss_stride_count_timestamp", -1);
            final float speed = intent.getFloatExtra("speed", -1.0f);
            final int cadence = intent.getIntExtra("cadence", -1);
            final int hr = intent.getIntExtra("hr", -1);
//            final float runSpeed = intent.getFloatExtra("ss_speed", -1);
//            final long runStrideCount = intent.getLongExtra("ss_stride_count", -1);
            final Instant instant = Instant.now();  // Current moment in UTC.
            LocalDateTime now = LocalDateTime.ofInstant(instant, ZoneId.systemDefault());

            runOnUiThread(new Runnable() {
                @SuppressLint("DefaultLocale")
                @Override
                public void run() {
                    if(statusBSD != null)
                        tv_speed.setText(statusBSD);
                    else
                    if (speed >= 0.0f)
                        tv_speed.setText(String.format("%.01f", speed) + " km/h");
                    if(statusBC != null)
                        tv_speed.setText(statusBC);
                    else
                    if (cadence >= 0)
                        tv_cadence.setText(String.format("%3d rpm", cadence));
                    if (now != null)
                        tv_time.setText(String.format("%2d:%02d:%02d",
                                now.getHour(), now.getMinute(), now.getSecond()));
                }
            });
        }
    }
}

File: app/src/main/java/idv/markkuo/cscblebridge/MainActivity.kt
package idv.markkuo.cscblebridge

import android.app.ActivityManager
import android.content.*
import android.os.Build
import android.os.Bundle
import android.os.IBinder
import android.util.Log
import android.widget.Button
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import idv.markkuo.cscblebridge.CSCService.LocalBinder
import java.time.Instant
import java.time.LocalDateTime
import java.time.ZoneId

class MainActivity : AppCompatActivity() {
    private val tv_speedSensorState: TextView? = null
    private val tv_cadenceSensorState: TextView? = null
    private val tv_hrSensorState: TextView? = null
    private val tv_runSensorState: TextView? = null
    private val tv_speedSensorTimestamp: TextView? = null
    private val tv_cadenceSensorTimestamp: TextView? = null
    private val tv_hrSensorTimestamp: TextView? = null
    private val tv_runSensorTimestamp: TextView? = null
    private var tv_speed: TextView? = null
    private var tv_cadence: TextView? = null
    private val tv_hr: TextView? = null
    private val tv_runSpeed: TextView? = null
    private val tv_runCadence: TextView? = null
    private var tv_time: TextView? = null
    private var btn_service: Button? = null
    private var receiver: MainActivityReceiver? = null
    private var mBound = false
    private var mService: CSCService? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val intent = intent
        val mServiceIntent = Intent(applicationContext, CSCService::class.java)
        setContentView(R.layout.activity_main)
        tv_speed = findViewById(R.id.SpeedText)
        tv_cadence = findViewById(R.id.CadenceText)
        tv_time = findViewById(R.id.TimeText)
        ensureServiceRunning(mServiceIntent)
        receiver = MainActivityReceiver()
        // register intent from our service
        val filter = IntentFilter()
        filter.addAction("idv.markkuo.cscblebridge.ANTDATA")
        registerReceiver(receiver, filter)
    }

    private fun ensureServiceRunning(mServiceIntent: Intent) {
        if (mBound) {
            Log.d(TAG, "Service already started")
        } else {
            Log.d(TAG, "Starting Service")
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) startForegroundService(mServiceIntent) else startService(mServiceIntent)

            // Bind to the service so we can interact with it
            if (!bindService(mServiceIntent, connection, Context.BIND_AUTO_CREATE)) {
                Log.d(TAG, "Failed to bind to service")
            } else {
                mBound = true
            }
        }
    }

    override fun onResume() {
        val mServiceIntent = Intent(applicationContext, CSCService::class.java)
        super.onResume()
        ensureServiceRunning(mServiceIntent)
    }

    /** Defines callbacks for service binding, passed to bindService()  */
    private val connection: ServiceConnection = object : ServiceConnection {
        override fun onServiceConnected(className: ComponentName, service: IBinder) {
            val binder = service as LocalBinder
            mService = binder.service
            mBound = true
        }

        override fun onServiceDisconnected(arg0: ComponentName) {
            mBound = false
        }
    }

    // Unbind from the service
    fun unbindService() {
        if (mBound) {
            unbindService(connection)
            mBound = false
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        unregisterReceiver(receiver)
        unbindService()
    }

    private fun resetUi() {
        tv_speed!!.text = getText(R.string.no_data)
        tv_cadence!!.text = getText(R.string.no_data)
        tv_time!!.text = "--:--"
    }

    private val isServiceRunning: Boolean
        private get() {
            val manager = (getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager)
            for (service in manager.getRunningServices(Int.MAX_VALUE)) {
                if (CSCService::class.java.name == service.service.className) {
                    return true
                }
            }
            return false
        }

    // this BroadcastReceiver is used to update UI only
    private inner class MainActivityReceiver : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            val statusBSD = intent.getStringExtra("bsd_service_status") // bicycle speed
            val statusBC = intent.getStringExtra("bc_service_status") // bicycle cadence
            //            final String statusHR = intent.getStringExtra("hr_service_status");
//            final String statusRsc = intent.getStringExtra("ss_service_status");
//            final long speedTimestamp = intent.getLongExtra("speed_timestamp", -1);
//            final long cadenceTimestamp = intent.getLongExtra("cadence_timestamp", -1);
//            final long hrTimestamp = intent.getLongExtra("hr_timestamp", -1);
//            final long runTimestamp = intent.getLongExtra("ss_stride_count_timestamp", -1);
            val speed = intent.getFloatExtra("speed", -1.0f)
            val cadence = intent.getIntExtra("cadence", -1)
            val hr = intent.getIntExtra("hr", -1)
            //            final float runSpeed = intent.getFloatExtra("ss_speed", -1);
//            final long runStrideCount = intent.getLongExtra("ss_stride_count", -1);
            val instant = Instant.now() // Current moment in UTC.
            val now = LocalDateTime.ofInstant(instant, ZoneId.systemDefault())
            runOnUiThread {
                if (statusBSD != null) tv_speed!!.text = statusBSD else if (speed >= 0.0f) tv_speed!!.text = String.format("%.01f", speed) + " km/h"
                if (statusBC != null) tv_speed!!.text = statusBC else if (cadence >= 0) tv_cadence!!.text = String.format("%3d rpm", cadence)
                if (now != null) tv_time!!.text = String.format("%2d:%02d:%02d",
                        now.hour, now.minute, now.second)
            }
        }
    }

    companion object {
        private val TAG = MainActivity::class.java.simpleName
    }
}